gcc/rust/rust-attribs.cc

```diff
@@ -217,9 +242,14 @@ handle_malloc_attribute (tree *node, tree, tree, int, bool *)
    struct attribute_spec.handler.  */
 
 static tree
-handle_pure_attribute (tree *node, tree, tree, int, bool *)
+handle_pure_attribute (tree *node, tree name, tree, int, bool *no_add_attrs)
 {
-  gcc_assert (TREE_CODE (*node) == FUNCTION_DECL);
+  if (TREE_CODE (*node) != FUNCTION_DECL)
+    {
+      warning (OPT_Wattributes, "%qE attribute ignored", name);
+      *no_add_attrs = true;
+    }
+
   DECL_PURE_P (*node) = 1;
   return NULL_TREE;
```

Comment:
`DECL_PURE_P` does `FUNCTION_DECL_CHECK`, therefore is valid for `FUNCTION_DECL` only, but you're now permitting reaching it for other DECLs, too.  Either needs an early `return NULL`, or otherwise restructuring of the code.

Also, what about adding test cases for the respective cases?  :-)